numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2674,16 +2654,30 @@ finish:
 #endif
     Py_DECREF(robj);
     return obj;
-#endif
 }
 #undef _@TYPE@_IS_@TYPE@
 
-#undef _NPY_UNUSED_TYPE
-#undef _NPY_UNUSED_KWDS
 /**end repeat**/
 
 #undef _WORK
 
+static PyObject *
+object_arrtype_new(PyTypeObject *NPY_UNUSED(type), PyObject *args, PyObject *kwds)
+{
+    PyObject *obj = Py_None;
+    char *kwnames[] = {"", NULL};  /* positional-only */
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "|O:object_", kwnames, &obj)) {
+        return NULL;
+    }
+    PyArray_Descr *typecode = PyArray_DescrFromType(NPY_OBJECT);
+    if (typecode == NULL) {
+        return NULL;
+    }
+    PyArrayObject *arr = (PyArrayObject *)PyArray_FromAny(obj, typecode,
+                                    0, 0, NPY_ARRAY_FORCECAST, NULL);
+    return PyArray_Return(arr);
```

Comment:
I feel like this can also be used in the above case. The `finish:` path really is only sensible in the case where `PyArray_FromAny` is not called.

Reply:
Probably true - I was struggling to piece together what was going on in the above case, so decided to split this part out first.

For some context - my eventual goal is to add a `Py_UCS4 *` member to `PyUnicodeScalarObject`, which needs to be populated during initialization - instead of using the deprecated `Py_UNICODE` APIs.